mindmeld/text_preparation/text_preparation_pipeline.py

```diff
@@ -0,0 +1,566 @@
+        """Static method to create a TextPreparationPipeline instance from an app_path.
+
+        Args:
+            app_path (str): The application path.
+
+        Returns:
+            TextPreparationPipeline: A TextPreparationPipeline class.
+        """
+        language, _ = get_language_config(app_path)
+        text_preparation_config = get_text_preparation_config(app_path)
+
+        if (
+            text_preparation_config.get("regex_norm_rules")
+            and "normalizers" not in text_preparation_config
+        ):
+            logger.warning(
+                "Detected 'regex_norm_rules' in TEXT_PREPARATION_CONFIG, however, 'normalizers' "
+                "have not been specified. Will apply specified 'regex_norm_rules' in addition to "
+                "default normalizers. To omit default normalizers set 'normalizers' to []."
+            )
+        normalizers = (
+            DEFAULT_NORMALIZERS
+            if "normalizers" not in text_preparation_config
+            else text_preparation_config.get("normalizers")
+        )
+        stemmer = (
+            "NoOpStemmer"
+            if "stemmer" in text_preparation_config
+            and not text_preparation_config["stemmer"]
+            else text_preparation_config.get("stemmer")
+        )
+        return TextPreparationPipelineFactory.create_text_preparation_pipeline(
+            language=language,
+            preprocessors=text_preparation_config.get("preprocessors"),
+            regex_norm_rules=text_preparation_config.get("regex_norm_rules"),
+            keep_special_chars=text_preparation_config.get("keep_special_chars"),
+            normalizers=normalizers,
+            tokenizer=text_preparation_config.get("tokenizer"),
+            stemmer=stemmer,
+        )
+
+    @staticmethod
+    def create_text_preparation_pipeline(
+        language: str = ENGLISH_LANGUAGE_CODE,
+        preprocessors: List[str] = None,
+        regex_norm_rules: List[Dict] = None,
+        keep_special_chars: str = None,
+        normalizers: List[str] = None,
+        tokenizer: Tokenizer = None,
+        stemmer: Stemmer = None,
+    ):
+        """Static method to create a TextPreparationPipeline instance.
+
+        Args:
+            language (str, optional): Language as specified using a 639-1/2 code.
+            preprocessors (List[str]): List of preprocessor class names.
+            regex_norm_rules (List[Dict]): List of regex normalization rules represented as
+                dictionaries. ({"pattern":<pattern>, "replacement":<replacement>})
+            normalizers (List[str]): List of normalizer class names.
+            tokenizer (str): Class name of Tokenizer to use.
+            stemmer (str): Class name of Stemmer to use.
+
+        Returns:
+            TextPreparationPipeline: A TextPreparationPipeline class.
+        """
+        preprocessors = (
+            [PreprocessorFactory.get_preprocessor(p) for p in preprocessors]
```

Comment:
I am not sure how a custom preprocessor will be used here since we only check for noop preprocessor in PreprocessorFactory

Reply:
So this is how it was being done previously

```
from mindmeld import Application
from .stem_processor import StemProcessor

preprocessor = StemProcessor()
app = Application(__name__, preprocessor=preprocessor)
```
https://www.mindmeld.com/docs/userguide/preprocessor.html

Instead of this you would place a 'text_preparation_pipeline' object. Which you would and then you can overwrite the preprocessor section with custom preprocessor objects.

Reply:
text_preparation_pipeline.preprocessors = [CustomPreprocessor()]
app = Application(__name__, text_preparation_pipeline=text_preparation_pipeline)

Reply:
Modified TextPreparationPipelineFactory.create_text_preparation_pipeline() to be able to accept either List[str] or List[Preprocessor] for each of the components as per our discussion